Add creatorId to rooms

avatarneil opened the pull request:
Fixes #80 

Adds a `creatorId` property to rooms. All users who join a room that are not the creator will no longer see an 'everyone is in button', but in the case that a user somehow _is_ still able to see that button, the BE will additionally check to make sure that the creatorId matches the socket id that is making the 'initiate swiping' request.

Additionally, introduces some newer javascript syntax, and refactors a few things to `const` as opposed to `let`. Mutability is gross!

avatarneil commented on react/src/components/Rooms/Lobby.js:
```diff
@@ -2,18 +2,32 @@ import React from 'react';
+
+
+const EveryoneIsInButton = ({isCreator, socket}) => {
+    if (isCreator) {
+        return <button onClick={() => startSwiping(socket)}> EVERYONE IS IN </button>
+    }
+    
+    return null
```
Could've just done a ternary for this and made it a oneliner as well ¯\_(ツ)_/¯

avatarneil commented:
Can we tweak the codacy rules? The flag about a destructure being an unnecessary block is a little bit ridiculous. Also, I've found most modern JS codebases typically prefer single-quote strings to double-quote, but happy to switch that over if desired.

kelvinfan001 commented on express/socket.js:
```diff
@@ -125,12 +125,17 @@ module.exports = (io) => {
+            const query = { roomNumber: socket.room, creatorId: socket.id };
+
+            try {
+                const result = await Room.findOneAndUpdate(query, { isActive: true }, { new: true })
```
haha yeah we should probably make everything use `await`. This is much cleaner, thanks!
But I think we forgot to make the function asynchronous ;) 
```
               socket.on("initiate swiping", async () => {
```

avatarneil replied:
Aha, yes! I'm spoiled by ESLint yelling at me to mark functions as `async` in other projects ^_^

kelvinfan001 reviewed (changes requested):
This is awesome!
Your code is definitely much cleaner than mine; really appreciate you cleaning up all my ugly syntax (that I might have copied and pasted from StackOverflow, I actually have limited experience with JavaScript).

One thing we should do before merging this, though, is make non-room-creator users get notified that the room creator disconnected (should probably also auto disconnect them from the room), since without the `EVERYONE IS IN` button, that room is pretty much dead anyway. It should be a simple fix, check out [this line](https://github.com/kelvinfan001/findalicious/blob/d158fd44a4e0ff145bbb1db3ef5cb63ef3a3571d/express/socket.js#L19). Instead of checking that the last participant left, let's just check that the room creator left, and do something like `io.sockets.in(roomNumber).emit('room creator disconnect', roomInfo);`, and in [this file](https://github.com/kelvinfan001/findalicious/blob/master/react/src/components/Rooms/Lobby.js), we add something like 
```
socket.on('room creator disconnect', (result) => {
    // alert users that room creator disconnected
    // redirect home
});
````

Alternatively, we could assign "room creator" status to some other user so we don't purge the whole room if the room creator experienced some network issues while still in the room lobby stage. I prefer the former option, though. WDYT?

kelvinfan001 commented:
Oh, and about the codacy stuff... Maybe we should just do away with it altogether, I also find it to be more of a nuisance. Feel free to just ignore it for now. And judging from this PR, I think you're probably more knowledgeable than me (or Codacy, for that matter) regarding modern JS conventions/rules.

kelvinfan001 commented:
@all-contributors please add avatarneil for code.

avatarneil commented:
Re: Codacy, in most projects I typically use a hybrid of eslint + prettier, which usually works really well. The VSCode integration is super rich, so it makes it really easy to catch style errors before you even push! We could totally hook that in with something like TravisCI, CircleCI, or whatever CI we wanna use. I'd love to cut a PR at some point introducing eslint :)

avatarneil commented:
Added the alerting & redirect to homescreen for when creators disconnect, as well as adding an async modifier to the method I added `await` to. Thanks for the pointers in your comment, incredibly helpful as I'm still trying to get cozy in the codebase!

kelvinfan001 commented:
I'm also mostly relying on the ESLint VSCode integration atm locally. But yeah, I think we should get rid of codacy. I'll look into some CI options we can use :)

kelvinfan001 reviewed (approved):
LGTM! 
Tested it out locally and is working great. Feel free to merge when ready.
